### PR TITLE
added time-based test logic and pbft test cases mechanisms

### PR DESCRIPTION
**Pull request by BatiGencho**
- added time-based validity checks
- added the ability to await all btc servers before running tests
- added a pbft e2e test (currently commented out)
- added the ability to signal the exact moment when signing has finished via session ids
- added better output and display traits masking long hex blocks
- reworked the time logic to be based on secs than on minutes and thus be more precise
- increased grpc timeout, ping and healthcheck durations
- added 2 new btc server grpc endpoints
- fixes to test-suite
- e2e tests renaming
- fmt, cleanup

### CODE REVIEW

File: bin/btc-server/proto/btc_server.proto [hunk 1]
```diff
@@ -159,3 +163,24 @@ message GetUTXOMerkleRootResponse {
...
+    uint32 max_results = 1;
+}
+
+message GetSessionIdsResponse {
+    repeated bytes data = 1;
+}
+
+enum SigningStatus {
+    RUNNING = 0;
+    FINALIZED = 1;
```
Comment (arminsabouri) [inline bin/btc-server/proto/btc_server.proto:185]: Should we have failed? 
If we a signing session for a epoch that expired then a different coord picked it up the prev one should be "failed"

File: bin/btc-server/src/cordinator.rs [hunk 2]
```diff
@@ -367,4 +368,38 @@ impl App {
...
+            self.db.get_psbt(signing_session_id)?.ok_or(CoordinatorError::CouldNotFindPsbt)?;
+
+        //psbt.finalize(secp)
+
+        let mut is_finalized = true;
+        for (_index, psbt_input) in psbt.inputs.iter().enumerate() {
+            if psbt_input.sighash_type.is_none() || psbt_input.tap_key_sig.is_none() {
+                is_finalized = false;
+                break;
+            }
```
Comment (arminsabouri) [inline bin/btc-server/src/cordinator.rs]: we talked about this in person but just for the record.
We know if signing was successful by calling finalize() (from the miniscript crate). Finalize checks the agg sig against each input. But in order for finalize to work each signer and coord needs to save the final PSBT with the witness. Im not sure if they persist the final psbt or if they just check it in the finalize endpoint.

File: bin/btc-server/src/database.rs [hunk 3]
```diff
@@ -433,6 +460,20 @@ mod tests {
...
+        let (db, _temp_dir) = setup_db();
+
+        let tx = create_tx(2);
+        let psbt = Psbt::from_unsigned_tx(tx).unwrap();
+        let signing_session_id: [u8; 32] = [0; 32];
+        db.update_psbt(&signing_session_id, &psbt).unwrap();
+        db.flush().unwrap();
+
+        let signing_session_ids = db.get_session_ids(10).unwrap();
+        println!("Signing session ids {:?}", signing_session_ids);
```
Comment (arminsabouri) [inline bin/btc-server/src/database.rs]: just a TODO for an assertion here

File: bin/btc-server/src/cordinator.rs (hunk 2 above)
Comment (BatiGencho) [reply bin/btc-server/src/cordinator.rs]: fixed

File: bin/btc-server/proto/btc_server.proto (hunk 1 above)
Comment (BatiGencho) [reply bin/btc-server/proto/btc_server.proto:185]: added

File: bin/btc-server/src/database.rs (hunk 3 above)
Comment (BatiGencho) [reply bin/btc-server/src/database.rs]: added

**Comment by arminsabouri** [comment]
Blocked
- Armin: Coin Selection error: Insufficient funds: sat available of sat needed keeps coming quite often and breaks the test suite. Shows up in e2e FROST test.
- Evgeni: dkg not finalizing. 
- Evgeni: Cleaning up logs

File: bin/btc-server/src/database.rs [hunk 4]
```diff
@@ -112,6 +114,41 @@ impl Db {
         }
     }
 
+    /// Get signing session ids from db
+    pub fn get_session_ids(&self, max_results: u32) -> Result<Vec<[u8; 32]>, Error> {
+        let mut ret = Vec::new();
+        let mut results = 0;
+        for res in self.psbt.iter() {
+            info!("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
```
Comment (arminsabouri) [inline bin/btc-server/src/database.rs]: rm this

File: bin/btc-server/src/database.rs [hunk 5]
```diff
@@ -112,6 +114,41 @@ impl Db {
...
+            let (k, _) = res?;
+            let signing_session_id: [u8; 32] =
+                k.to_vec().as_slice().try_into().map_err(Error::Serialization)?;
+            results += 1;
+            info!("XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX {:?}", signing_session_id);
```
Comment (arminsabouri) [inline bin/btc-server/src/database.rs]: samething here

File: bin/test-suite/src/suite/consensus/common/events.rs [hunk 6]
```diff
@@ -31,6 +32,25 @@ pub async fn await_dkg(
...
 
+pub async fn await_signing_completion(
+    in_turn_member_index: u16,
+    rx: &mut tokio::sync::mpsc::Receiver<Notifications>,
+) {
+    while let Some(notification) = rx.recv().await {
+        if let Notifications::SigningStatusReport((member_index, _session_id, status)) =
+            notification
+        {
+            info!(
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/common/events.rs]: Could remove this

File: bin/test-suite/src/suite/consensus/mod.rs [hunk 7]
```diff
@@ -60,22 +60,23 @@ impl Suite for ConsensusIntegrationTestSuite {
         self.set_panic_hook();
 
         // dkg tests
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/mod.rs:55]: uncomment tests

File: crates/consensus/authority/src/pbft.rs [hunk 8]
```diff
@@ -467,6 +554,13 @@ where
         if peer_id == self.peer_id {
             return Ok(None);
         }
+
+        // perform block validation
+        // if !self.validate_block(&block).await? {
+        //     warn!(target: "pbft" ,"Block proposal failed validation");
+        //     return Err(Error::BlockValidationFailed);
+        // }
```
Comment (arminsabouri) [inline crates/consensus/authority/src/pbft.rs]: Do we want to keep these?

File: crates/consensus/authority/src/pbft.rs [hunk 9]
```diff
@@ -243,10 +248,79 @@ where
         }
     }
 
+    async fn validate_block(&self, block: &SealedBlock) -> Result<bool, Error> {
```
Comment (arminsabouri) [inline crates/consensus/authority/src/pbft.rs]: Is this used? Any plans to use if not?

**Review by arminsabouri — CHANGES_REQUESTED** [review]

File: bin/btc-server/src/database.rs (hunk 4 above)
Comment (BatiGencho) [reply bin/btc-server/src/database.rs]: done

File: bin/btc-server/src/database.rs (hunk 5 above)
Comment (BatiGencho) [reply bin/btc-server/src/database.rs]: done

File: crates/consensus/authority/src/pbft.rs (hunk 8 above)
Comment (BatiGencho) [reply crates/consensus/authority/src/pbft.rs]: yes, I have reactivated the checks

File: crates/consensus/authority/src/pbft.rs (hunk 9 above)
Comment (BatiGencho) [reply crates/consensus/authority/src/pbft.rs]: it is being used, indeed

File: crates/consensus/authority/src/pbft.rs [hunk 10]
```diff
@@ -243,10 +248,79 @@ where
...
+            );
+            return Ok(false);
+        }
+        if block.timestamp <= start - BLOCK_TIME_VALIDITY_CRITERIA_RELAXATION.as_secs() {
+            error!(
+                "Block validity start range violation. Timestamp = {:?}. Range end = {:?}",
+                block.timestamp, start
+            );
+            return Ok(false);
+        }
```
Comment (arminsabouri) [inline crates/consensus/authority/src/pbft.rs]: This check is really similar to the is_inturn check we do can we just re-use that here?

Comment (arminsabouri) [reply crates/consensus/authority/src/pbft.rs]: Additionally I've added some tighter bounds around the timestamp that can proposed by a peer in https://github.com/botanix-labs/Macbeth/pull/191

File: crates/consensus/authority/src/pbft.rs [hunk 11]
```diff
@@ -243,10 +248,79 @@ where
         }
     }
 
+    async fn validate_block(&self, block: &SealedBlock) -> Result<bool, Error> {
+        // get current coordinator and its validity range
+        let coordinator_index =
+            self.get_coordinator().await?.map(|(_, _, authority_index)| authority_index);
+        let (start, end, _, _) = self.get_inturn_interval_for_coordinator(coordinator_index);
+
+        if block.timestamp >= end + BLOCK_TIME_VALIDITY_CRITERIA_RELAXATION.as_secs() {
```
Comment (arminsabouri) [inline crates/consensus/authority/src/pbft.rs]: b/c of the relaxation time, is there a possibility that two blocks produced by two peers are valid for the same minute time slot?

File: bin/test-suite/src/suite/consensus/pbft/test_pbft_disconnect.rs [hunk 12]
```diff
@@ -191,83 +194,46 @@ pub async fn pbft_e2e_stable(
...
+    // wait for a few seconds before disconnecting a peer
+    tokio::time::sleep(Duration::from_secs(3)).await;
 
-    // wait for the tx to be included in a botanix block
-    await_botanix_event(&mut rx, *BURN_TOPIC).await;
+    // now disconnect the peers of another peer that is not a coordinator
+    let next_peer_index = (inturn_member_index + 1).max(total_authorities as u64 - 1);
 
-    // make sure we have enough time for the nonce to be updated
+    test_fed_members
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/pbft/test_pbft_disconnect.rs:210]: What event is suppose to happen in the 10 seconds that this peer's peers are disconnected?

File: crates/primitives/src/header_ext.rs [hunk 13]
```diff
@@ -184,8 +184,15 @@ impl HeaderExt for Header {
...
-        if (block_timestamp_min / authorities_len) % authorities_len != (signer_index as u64) {
+        let cycle_length = authorities_len * 60; // Define the cycle length in seconds
+        let block_timestamp_sec = self.timestamp; // Use the block timestamp in seconds
+
+        // Calculate the current cycle's position and determine the current in-turn signer
+        let current_cycle_position = block_timestamp_sec % cycle_length;
+        let current_in_turn_index = (current_cycle_position / 60) % authorities_len;
+
+        // Check if the calculated index matches the signer index
+        if current_in_turn_index != (signer_index as u64) {
```
Comment (arminsabouri) [inline crates/primitives/src/header_ext.rs:195]: Can we re-use the method defined in utils here
https://github.com/botanix-labs/Macbeth/pull/190/files#diff-70d823804b6ce466fbab201283c5261894ba0ba35696b2902c9abeced5815ab4R268

They seem to be doing the samething

**Review by arminsabouri — COMMENTED** [review]
Two small questions

File: crates/consensus/authority/src/pbft.rs [hunk 14]
```diff
@@ -1467,7 +1478,7 @@ mod tests {
             let pk = sk.public_key(secp256k1::SECP256K1);
             let mut recovered = false;
             for sig in sigs_so_far.iter() {
-                let msg = secp256k1::Message::from_slice(
+                let msg = secp256k1::Message::from_digest_slice(
```
Comment (arminsabouri) [inline crates/consensus/authority/src/pbft.rs:1486]: What was the reasoning behind this change?

Comment (BatiGencho) [reply crates/consensus/authority/src/pbft.rs:1486]: The linter suggested this as the from_slice is already deprecated.

File: bin/test-suite/src/suite/consensus/pbft/test_pbft_disconnect.rs (hunk 12 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/pbft/test_pbft_disconnect.rs:210]: I am expecting a pbft restart

File: crates/primitives/src/header_ext.rs (hunk 13 above)
Comment (BatiGencho) [reply crates/primitives/src/header_ext.rs:195]: This creates a circular dependency. The edh header rs is inside hte primitives crate  which the consensus is using. We cannot import the consensus back into it.

**Pull request merged**